impl/sql/mutationstorage/queue.go

```diff
@@ -125,37 +140,36 @@ func (m *Mutations) send(ctx context.Context, domainID string, logID int64, mDat
 	return tx.Commit()
 }
 
-// HighWatermarks returns the highest timestamp for each log in the mutations table.
-func (m *Mutations) HighWatermarks(ctx context.Context, domainID string) (map[int64]int64, error) {
-	watermarks := make(map[int64]int64)
-	rows, err := m.db.QueryContext(ctx,
-		`SELECT LogID, Max(Time) FROM Queue WHERE DomainID = ? GROUP BY LogID;`,
-		domainID)
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
-	for rows.Next() {
-		var logID, watermark int64
-		if err := rows.Scan(&logID, &watermark); err != nil {
-			return nil, err
-		}
-		watermarks[logID] = watermark
-	}
-	if err := rows.Err(); err != nil {
-		return nil, err
-	}
-	return watermarks, nil
+// HighWatermark returns the highest watermark in logID that is less than or
+// equal to batchSize items greater than start.
+func (m *Mutations) HighWatermark(ctx context.Context, domainID string, logID,
+	start int64, batchSize int32) (int32, int64, error) {
+	var count int32
+	var high int64
+	if err := m.db.QueryRowContext(ctx,
+		`SELECT COUNT(*), COALESCE(MAX(T1.Time), ?) FROM 
+		(
+			SELECT Q.Time FROM Queue as Q
+			WHERE Q.DomainID = ? AND Q.LogID = ? AND Q.Time > ?
```

Comment:
Wait, do you store time? I'm not sure MySQL time is reliably ordered. Would auto_increment key be better?

The assumption that **must** be strongly supported by your storage is that once watermark == x, there will be no new messages with a lower value. I don't think it works with time just the way it works in Spanner.

Reply:
Ah, found in the code that you query for max existing timestamp before creating a new one, so it should be correct. I'm not sure though how this would compare to auto_increment key in performance, perhaps needs some testing.

Reply:
On the third hand, if you already read max(Timestamp) in a transaction then why can't it be just LeafIndex instead of Timestamp? You would then read it, +1, and write.